### PR TITLE
Caching meshes per 3d viewer and handle adding instances only once to scene

### PR DESCRIPTION
**Pull request by ddelpiano**
- Meshes are cached in the MeshFactory so that we are not re-computing them all the times, this works per 3d viewer instance.
- webgl scene not purged all the time
- instances already presents are edited (color and opacity) if required
- only new instances are added to the scene

### CODE REVIEW

File: geppetto.js/geppetto-ui/src/3d-canvas/Canvas.js [hunk 1]
```diff
@@ -62,7 +62,7 @@ class Canvas extends Component {
 
     if (prevProps !== this.props){
       const { data, cameraOptions, threeDObjects } = this.props;
-      await this.threeDEngine.update(data, cameraOptions, threeDObjects, this.shouldEngineTraverse());
+      await this.threeDEngine.update(data, cameraOptions, threeDObjects, this.shouldEngineTraverse(data, threeDObjects));
```
Comment (ddelpiano) [inline geppetto.js/geppetto-ui/src/3d-canvas/Canvas.js:65]: note for myself, to clean up this.

File: geppetto.js/geppetto-ui/src/3d-canvas/threeDEngine/ThreeDEngine.js [hunk 2]
```diff
@@ -198,17 +198,141 @@ export default class ThreeDEngine {
...
+
+    if (instances) {
+      topInstances = instances.map((i) => { return i.instancePath });
+      for (var i = toRemove.length - 1; i >= 0; i--) {
+        for (var j = 0; j <= topInstances.length - 1; j++) {
+          // if the Mesh is present in the list of instances provided by the canvas then
+          // we can remove this from that list itself and from the toRemove so that we don't recompute
+          // the same mesh/instance twice
+          if (toRemove[i].instancePath.startsWith(topInstances[j])) {
+            // if the color is the same then remove this
```
Comment (ddelpiano) [inline geppetto.js/geppetto-ui/src/3d-canvas/threeDEngine/ThreeDEngine.js:315]: if the color is the same we don't update, differently we do that and the remove the instance. Comment to be updated

Comment (afonsobspinto) [reply geppetto.js/geppetto-ui/src/3d-canvas/threeDEngine/ThreeDEngine.js:315]: I'm not a big fan of having a method called clearScene do more than just clearing the scene according to our criteria. (aka I don't like that it also changes colors besides removing stuff)
But I understand the need of having it there.

The first idea that I had was to not remove the instance from instances in this method but rather add some attribute to it so that we could skip the meshFactory work based on that attribute.
But that's just an idea.

File: geppetto.js/geppetto-ui/src/3d-canvas/threeDEngine/TrackballControls.js [hunk 3]
```diff
@@ -143,9 +143,6 @@ THREE.TrackballControls = function (object, domElement, handler) {
   };
 
   this.setCameraByConsole = function () {
-    if (_this.cameraByConsoleLock) {
```
Comment (afonsobspinto) [inline geppetto.js/geppetto-ui/src/3d-canvas/threeDEngine/TrackballControls.js:146]: What's the reason behind this change? In #175 I ported TrackballControls to the latest version so whatever we change here now, needs to be ported again in #175

File: geppetto.js/geppetto-ui/src/3d-canvas/threeDEngine/ThreeDEngine.js [hunk 4]
```diff
@@ -198,17 +198,141 @@ export default class ThreeDEngine {
...
+          if (child.instancePath === mesh.instancePath && child.uuid === mesh.uuid) {
+            child.material.color.setRGB(instance.color.r, instance.color.b, instance.color.b);
+            if (instance.color.a) {
+              child.material.transparent = true;
+              child.material.opacity = instance.color.a;
+            }
+          }
+        }
+      } else {
+        console.error("color for instance provided is malformed, please provide the color in the form");
```
Comment (afonsobspinto) [inline geppetto.js/geppetto-ui/src/3d-canvas/threeDEngine/ThreeDEngine.js:270]: I don't think instances **have** to have a color, it can be undefined and we add a default color

File: geppetto.js/geppetto-ui/src/3d-canvas/threeDEngine/MeshFactory.js [hunk 5]
```diff
@@ -270,13 +269,39 @@ export default class MeshFactory {
...
+        && currentInstance?.material?.color?.g === material?.color?.g
+        && currentInstance?.material?.color?.b === material?.color?.b
+        && currentInstance?.material?.color?.opacity === material?.color?.a) {
+          return true;
+        } else {
+          return false;
+        }
+    } else if (currentInstance?.type === 'Group') {
+      for (let child of currentInstance?.children) {
+        this.checkMaterial(child, material);
```
Comment (afonsobspinto) [inline geppetto.js/geppetto-ui/src/3d-canvas/threeDEngine/MeshFactory.js:290]: Not using the return value?
